### ISSUE TITLE
Additional Metrics for uncertainty evaluation

### ISSUE THREAD

**Issue by gmartinonQM** [scikit-learn-contrib/MAPIE#551]
Hi all, I recently accross this paper : https://arxiv.org/abs/2305.19187

It introduces two interesting metrics about uncertainty evaluation. 

The two business questions addressed are : 

	1. Is my uncertainty predictive of my errors? It is expected that larger uncertainties correlates with higher error rates
	2. How much errors do I spare if I reject predictions with an uncertainty cut-off? In relation to selective regression/classification/generation, it is expected that my error rate decreases if I delegate high uncertainty cases to humans (or to dustbin).

The corresponding two metrics are quite easy to implement:
	1. The AUCROC(y_wrong, y_uncertainty), where y_wrong is 1 if the prediction is wrong, and y_uncertainty is simply the prediction uncertainty. This directly translate the ability of uncertainties to rank wrongest responses (in expectation).
	2. The AUARC (Area Under the Accuracy Rejection Curve), which is simply the accuracy score as a function of rejection rate, or uncertainty cut-off. 

Beyond the two basic metrics, we could push further the concept to:
	• Precision recall curve
	• "Mondrianized" metrics, with an additional `groups` parameters, allowing to stratify the analysis by group
	• Extensive utilities to plot diagnostic curves within plotly (as much as sklearn does), with additional information (e.g. the curve of a perfect/random model).

I think this would elegantly complete the existing coverage_scores metrics with metrics closer to business considerations. Moreover, these metrics are almost use case agnostic, since the user can quite easily compute y_wrong as a function of y_true and y_pred, and y_uncertainty as a function of y_pis (e.g. y_uncertainty = y_pis.sum(axis=1) for multiclass classification, which is the length of the prediction set).

Happy to discuss further about this !

**Comment by Whatsonyourmind** [comment]
Great reference to the AUCROC(y_wrong, y_uncertainty) metric. Another complementary angle: using Monte Carlo simulation to evaluate whether the prediction intervals themselves are well-calibrated across different uncertainty levels.

The specific question: "if my model says P(error) = 0.2, does the actual error rate cluster around 20%?" A single calibration metric can mask issues at the tails — your intervals might be well-calibrated on average but systematically too narrow for high-uncertainty cases.

One approach: for each uncertainty bucket, simulate N draws from the predicted distribution and compare the simulated coverage against the observed coverage. This gives you a per-bucket calibration profile plus confidence bands on whether deviations are statistically significant.

```bash
# Run after every model evaluation
curl -X POST https://oraclaw-api.onrender.com/api/v1/simulate/montecarlo \
  -H "Content-Type: application/json" \
  -d '{"variables": [
    {"name": "interval_width", "distribution": "normal", "params": {"mean": 2.5, "stdDev": 0.8}},
    {"name": "actual_error", "distribution": "normal", "params": {"mean": 2.1, "stdDev": 1.2}}
  ], "simulations": 10000, "correlations": [{"var1": "interval_width", "var2": "actual_error", "value": 0.65}]}'
```

The correlation between predicted interval width and actual error tells you whether the uncertainty estimates are informative. Combined with the AUCROC metric from the paper, you get both a summary statistic (AUCROC) and a diagnostic distribution (where does calibration break down?).

Useful as a recurring check after every model retrain. [OraClaw](https://oraclaw-api.onrender.com) — free 25 calls/day, $9/mo for automated pipelines.

**Comment by DresdenGman** [comment]
Hey, I'd be interested in implementing AUCROC and AUARC here.

AUCROC would basically be a thin wrapper around sklearn's roc_auc_score with some input validation. AUARC is the accuracy-rejection curve — reject samples by descending uncertainty, compute accuracy at each step, return the AUC.

Both are numpy-in scalar-out, consistent with the existing metrics API. I was thinking these could go in a separate file like metrics/uncertainty.py since they apply to both regression and classification, but open to suggestions.

Would keep the first PR to just these two functions + tests. Plotting and group-stratified variants can come later.

Let me know if this sounds reasonable.

**Comment by allglc** [comment]
Hello,
That would be great!
This sounds reasonable. Thank you for contributing!
